### PR TITLE
Improve Code Quality

### PR DESCRIPTION
**Pull request by iamcarbon**
- Make various small classes structs
- Pass various structs by readonly ref (in)
- Eliminate various substring allocations
- Eliminate various virtual calls
- Optimize ReadHelper methods
- Optimize NameTokenizer

### CODE REVIEW

**Comment by iamcarbon** [comment]
@BobLd Ready for review / feedback.

**Comment by BobLd** [comment]
@iamcarbon can you include my PR changes from https://github.com/UglyToad/PdfPig/pull/817/commits/f7d1aee5ed73362cacfb25cf5013ec64bf639957 in this PR.

I want to avoid having to fix conflicts later on

**Review by BobLd — CHANGES_REQUESTED** [review]

File: src/UglyToad.PdfPig.Core/ReadHelper.cs [hunk 1]
```diff
@@ -41,7 +41,7 @@ public static class ReadHelper
             '\f'
         ];
 
-        private static readonly int MaximumNumberStringLength = long.MaxValue.ToString("D").Length;
+        private const int MaximumNumberStringLength = 20; // max length of a formatted Int64 value
```
[the remark this answers is not in the document]
Comment (BobLd) [reply src/UglyToad.PdfPig.Core/ReadHelper.cs:44]: Unless I'm mistaken, it's only positive numbers.

[the remark this answers is not in the document]
Comment (iamcarbon) [reply src/UglyToad.PdfPig.Core/ReadHelper.cs:44]: Yep, these should always be unsigned. I split these max values based on whether we're reading an Int32 or Int64 value to the max unsigned length, and eliminated the additional byte[] allocation here.

**Comment by iamcarbon** [comment]
> @iamcarbon can you include my PR changes from [f7d1aee](https://github.com/UglyToad/PdfPig/commit/f7d1aee5ed73362cacfb25cf5013ec64bf639957) in this PR.
> 
> I want to avoid having to fix conflicts later on

Included!

**Review by BobLd — APPROVED** [review]

**Pull request merged**